File: src/main/java/org/simbrain/network/dl4j/NeuronArray.java
package org.simbrain.network.dl4j;


import org.jetbrains.annotations.NotNull;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.factory.Nd4j;
import org.simbrain.network.LocatableModel;
import org.simbrain.network.core.Network;
import org.simbrain.network.events.NeuronArrayEvents;
import org.simbrain.util.UserParameter;
import org.simbrain.util.Utils;
import org.simbrain.util.propertyeditor.EditableObject;
import org.simbrain.workspace.AttributeContainer;
import org.simbrain.workspace.Consumable;
import org.simbrain.workspace.Producible;

import java.awt.geom.Point2D;
import java.awt.geom.Rectangle2D;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * High performance immutable array backed by ND4J Array.
 */
public class NeuronArray implements EditableObject, AttributeContainer, ArrayConnectable, LocatableModel {

    //TODO: Rename ideas: Array, Layer, ND4J Array, Double Array
    //TODO: See if data can be stored as an array. If not maybe used column instead of row.

    /**
     * Reference to network this array is part of.
     */
    private final Network parent;

    /**
     * A label for this Neuron Array for display purpose.
     */
    @UserParameter(label = "Label")
    private String label = "";

    /**
     * Id of this array.
     */
    @UserParameter(label = "ID", description = "Id of this array", order = -1, editable = false)
    private final String id;

    /**
     * Number of columns in the under laying ND4J Array.
     */
    @UserParameter(label = "Nodes", description = "Number of nodes", editable = false, order = 1)
    private int numNodes = 100;

    @UserParameter(label = "Increment amount", increment = .1, order = 20)
    private double increment = .1;

    /**
     * ND4J Array backing this object
     */
    private INDArray neuronArray;

    /**
     * For buffered update.
     */
    private INDArray arrayBuffer;

    /**
     * Center of the neuron array.
     */
    private double x;

    /**
     * Center of the neuron array.
     */
    private double y;

    /**
     * Reference to incoming weight matrix. TODO: Consider making this a list, prior to implementing serialization. But
     * note that there is no support currently for many-to-one connections.
     */
    private WeightMatrix incomingWeightMatrix;

    /**
     * "Fan-out" of outgoing weight matrices.
     */
    private List<WeightMatrix> outgoingWeightMatrices = new ArrayList<>();

    /**
     * Render an image showing each activation when true.
     */
    @UserParameter(label = "Show activations", description = "Whether to show activations as a pixel image", order = 4)
    private boolean renderActivations = true;

    /**
     * Event support.
     */
    private transient NeuronArrayEvents events = new NeuronArrayEvents(this);

    /**
     * Construct a neuron array.
     *
     * @param net      parent net
     * @param numNodes number of nodes
     */
    public NeuronArray(Network net, int numNodes) {
        parent = net;
        this.id = net.getIdManager().getId(NeuronArray.class);
        this.numNodes = numNodes;
        randomize();
    }

    /**
     * Make a deep copy of this array.
     *
     * @param newParent the new parent network
     * @param orig      the array to copy
     * @return the deep copy
     */
    public NeuronArray deepCopy(Network newParent, NeuronArray orig) {
        NeuronArray copy = new NeuronArray(newParent, orig.getNumNodes());
        copy.setLabel(copy.getId());
        copy.x = orig.x;
        copy.y = orig.y;
        copy.setValues(orig.getValues());
        return copy;
    }

    @Consumable()
    public void setValues(double[] values) {
        float[] floatValues = Utils.castToFloat(values);
        // TODO: Deal with the float issue, and make this set values in place of a float array, rather than
        //  re-creating it every time! However there may be an nd4j optimization such that under the hood
        //  not much object creation is happening
        neuronArray = Nd4j.create(floatValues).reshape(neuronArray.rows(), neuronArray.columns());
    }

    @Producible()
    public double[] getValues() {
        return Nd4j.toFlattened(neuronArray).toDoubleVector();
    }

    /**
     * Set the label. This prevents the group id being used as the label for new groups.  If null or empty labels are
     * sent in then the group label is used.
     */
    @Consumable(defaultVisibility = false)
    public void setLabel(String label) {
        String oldLabel = this.label;
        this.label = label;
        events.fireLabelChange(oldLabel, label);
    }

    /**
     * Simple randomization for now.
     */
    public void randomize() {
        neuronArray = Nd4j.rand(1, numNodes).subi(0.5).mul(2); // row vector
        events.fireUpdated();
    }

    public void update() {

        // TODO: This is just a place holder. Do something useful.
        // neuronArray = Nd4j.rand(10,10).subi(0.5).mul(2);

        events.fireUpdated();
    }

    public int getNumNodes() {
        return (int) neuronArray.length();
    }

    public INDArray getNeuronArray() {
        return neuronArray;
    }

    @NotNull
    @Override
    public Point2D getLocation() {
        return new Point2D.Double(x, y);
    }

    @Override
    public void setLocation(Point2D location) {
        this.x = location.getX();
        this.y = location.getY();
        fireLocationChange();
    }

    @Override
    public Rectangle2D getBound() {
        return new Rectangle2D.Double(x - 150 / 2, y - 50 / 2, 150, 50);
    }

    public boolean isRenderActivations() {
        return renderActivations;
    }

    public void setRenderActivations(boolean renderActivations) {
        this.renderActivations = renderActivations;
    }

    @Override
    public WeightMatrix getIncomingWeightMatrix() {
        return incomingWeightMatrix;
    }

    public void setIncomingWeightMatrix(WeightMatrix incomingWeightMatrix) {
        this.incomingWeightMatrix = incomingWeightMatrix;
    }

    @Override
    public List<WeightMatrix> getOutgoingWeightMatrices() {
        return outgoingWeightMatrices;
    }

    @Override
    public void addOutgoingWeightMatrix(WeightMatrix outgoingWeightMatrix) {
        this.outgoingWeightMatrices.add(outgoingWeightMatrix);
    }

    @Override
    public void removeOutgoingWeightMatrix(WeightMatrix weightMatrix) {
        this.outgoingWeightMatrices.remove(weightMatrix);
    }

    @Override
    public void onCommit() {
        events.fireLabelChange("", label);
    }

    public String getLabel() {
        return label;
    }

    @Override
    public String getName() {
        return "Neuron Array";
    }

    /**
     * Offset this neuron array
     *
     * @param offsetX x offset for translation.
     * @param offsetY y offset for translation.
     */
    public void offset(final double offsetX, final double offsetY) {
        x += offsetX;
        y += offsetY;
        events.fireUpdated();
    }

    /**
     * Add increment to every entry in weight matrix
     */
    public void increment() {
        neuronArray.addi(increment);
        events.fireUpdated();
    }

    /**
     * Subtract increment from every entry in weight matrix
     */
    public void decrement() {
        neuronArray.subi(increment);
        events.fireUpdated();
    }

    /**
     * Clear activations;
     */
    public void clear() {
        neuronArray.assign(0);
        events.fireUpdated();
    }

    /**
     * Since Neuron Array is immutable, this object will be used in the creation dialog.
     */
    public static class CreationTemplate implements EditableObject {

        /**
         * Number of columns in the under laying ND4J Array.
         */
        @UserParameter(label = "Nodes", description = "Number of nodes", order = 1)
        private int numNodes = 100;

        /**
         * A label for this Neuron Array for display purpose.
         */
        @UserParameter(label = "Label", description = "If left blank, a default label will be created.", initialValueMethod = "getLabel")
        private String label;

        /**
         * Create the template with a proposed label
         */
        public CreationTemplate(String proposedLabel) {
            this.label = proposedLabel;
        }

        /**
         * Add a neuron array to network created from field values which should be setup by an Annotated Property
         * Editor.
         *
         * @param network the network this neuron array adds to
         * @return the created neuron array
         */
        public NeuronArray create(Network network) {
            NeuronArray na = new NeuronArray(network, numNodes);
            na.label = label;
            return na;
        }

        /**
         * Getter called by reflection by {@link UserParameter#initialValueMethod}
         */
        public String getLabel() {
            return label;
        }

        @Override
        public String getName() {
            return "Neuron Array";
        }
    }

    @Override
    public INDArray getOutputArray() {
        return neuronArray;
    }

    @Override
    public long inputSize() {
        return neuronArray.length();
    }

    @Override
    public long outputSize() {
        return neuronArray.length();
    }

    @Override
    public void setInputArray(INDArray activations) {
        neuronArray = activations;
        numNodes = activations.columns();
    }

    @Override
    public void setInputBuffer(INDArray activations) {
        arrayBuffer = activations;
    }

    @Override
    public void setBufferValues() {
        // Not needed; Weight matrix does this.
    }

    @Override
    public void applyBufferValues() {
        if (arrayBuffer != null) {
            setInputArray(arrayBuffer.dup());
        }
    }


    @Override
    public String getId() {
        return id;
    }

    public void fireLocationChange() {
        events.fireLocationChange();
    }

    @Override
    public void onLocationChange(Runnable task) {
        events.onLocationChange(task);
    }

    @Override
    public Network getNetwork() {
        return parent;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("Array [" + getId() + "] with " + inputSize() + " components\n");
        // TODO: For larger numbers could present as a matrix
        int maxToDisplay = 10;
        if (neuronArray.length() < maxToDisplay) {
            sb.append(Arrays.toString(neuronArray.toDoubleVector()));
        }
        return sb.toString();
    }

    public NeuronArrayEvents getEvents() {
        return events;
    }

    @Override
    public void postUnmarshallingInit() {
        if (events == null) {
            events = new NeuronArrayEvents(this);
        }
    }

}

File: src/main/kotlin/org/simbrain/network/core/NeuronArray.kt
package org.simbrain.network.core

import org.simbrain.network.events.NeuronArrayEvents
import org.simbrain.network.updaterules.LinearRule
import org.simbrain.network.updaterules.NeuronUpdateRule
import org.simbrain.network.util.MatrixDataHolder
import org.simbrain.network.util.ScalarDataHolder
import org.simbrain.network.util.SpikingMatrixData
import org.simbrain.util.*
import org.simbrain.util.math.SimbrainMath
import org.simbrain.util.propertyeditor.EditableObject
import org.simbrain.util.propertyeditor.GuiEditable
import org.simbrain.util.stats.ProbabilityDistribution
import org.simbrain.workspace.AttributeContainer
import org.simbrain.workspace.Consumable
import org.simbrain.workspace.Producible
import smile.math.matrix.Matrix
import smile.stat.distribution.GaussianDistribution

/**
 * A "neuron array" backed by a Smile Matrix. Stored as a column vector.
 */
class NeuronArray(inputSize: Int) : ArrayLayer(inputSize), EditableObject, AttributeContainer {

    var updateRule: NeuronUpdateRule<ScalarDataHolder, MatrixDataHolder> by GuiEditable(
        initValue = LinearRule(),
        order = 100,
        typeMapProvider = NeuronUpdateRule<*, *>::getNeuronArrayTypeMap,
        setter = {
            val typeChanged = field::class != it::class
            field = it
            if (typeChanged) {
                dataHolder = updateRule.createMatrixData(size)
            }
            events.updated.fire()
        }
    )

    /**
     * Holds data for prototype rule.
     */
    var dataHolder: MatrixDataHolder by GuiEditable(
        initValue = updateRule.createMatrixData(inputSize),
        order = 99,
        onUpdate = {
            val proposedDataHolder = widgetValue(::updateRule).createMatrixData(size)
            if (widgetValue(::dataHolder)::class != proposedDataHolder::class) {
                refreshValue(proposedDataHolder)
            }
        }
    )

    /**
     * Array to hold activation values. These are also the outputs that are consumed by
     * other network components via [Layer]. A column vector.
     */
    @UserParameter(label = "Activations", description = "Neuron activations", order = 1)
    @get:Producible
    override var activations: Matrix = Matrix(inputSize, 1)
        set(newActivations) {
            field.copyFrom(newActivations)
            events.updated.fire()
        }

    @get:Producible
    @UserParameter("Bias Array", "Biases", order = 10)
    override var biases: Matrix = Matrix(inputSize, 1)
        set(newBiases) {
            field.copyFrom(newBiases)
            events.updated.fire()
        }

    @get:Producible
    override val biasArray: DoubleArray
        get() = biases.toDoubleArray()

    /**
     * see [AbstractNeuronCollection.spikes]
     */
    @get:Producible
    override val spikes: DoubleArray
        get() = (dataHolder as? SpikingMatrixData)?.spikes?.map { if (it) 1.0 else 0.0 }?.toDoubleArray() ?: DoubleArray(
            size
        )

    private var targets: Matrix? = null

    /**
     * Render an image showing each activation when true.
     */
    @UserParameter(label = "Show activations", description = "Whether to show activations as a pixel image", order = 4)
    var isRenderActivations = true

    @UserParameter(
        label = "Grid Mode", description = "If true, show activations as a grid, " +
                "otherwise show them as a line", order = 10
    )
    var gridMode = false
        set(value) {
            field = value
            (events as NeuronArrayEvents).visualPropertiesChanged.fire()
        }

    @UserParameter(
        label = "Vertical Layout",
        description = "If true, orient the array vertically, otherwise horizontally",
        order = 11
    )
    var verticalLayout = false
        set(value) {
            field = value
            (events as NeuronArrayEvents).visualPropertiesChanged.fire()
        }

    @UserParameter(label = "Biases Visible", description = "If true, show biases.", order = 12)
    var isShowBias = false
        set(showBias) {
            field = showBias
            events.visualPropertiesChanged.fire()
        }

    @Transient
    override var events: NeuronArrayEvents = NeuronArrayEvents()

    /**
     * Construct a neuron array.
     *
     * @param net  parent net
     * @param size number of components in the array
     */
    init {
        randomize()
    }

    /**
     * Make a deep copy of this array.
     *
     * @param newParent the new parent network
     * @return the deep copy
     */
    fun copy(): NeuronArray {
        val copy = NeuronArray(size)
        copy.location = location
        copy.gridMode = gridMode
        copy.verticalLayout = verticalLayout
        copy.activations.copyFrom(activations)
        copy.biases.copyFrom(biases)
        copy.updateRule = updateRule
        copy.dataHolder = dataHolder.copy()
        return copy
    }

    @get:Producible(arrayDescriptionMethod = "getLabelArray")
    override val activationArray: DoubleArray
        get() = activations.toDoubleArray()

    var targetValues: Matrix?
        get() = targets
        set(targets) {
            if (targets != null) {
                activations.validateSameShape(targets)
            }
            this.targets = targets
            events.updated.fire()
        }

    override fun randomize(randomizer: ProbabilityDistribution?) {
        activations = Matrix.rand(
            size, 1,
            GaussianDistribution(0.0, 1.0)
        )
        events.updated.fire()
    }

    override val name: String
        get() = "Neuron Array"

    /**
     * Offset this neuron array
     *
     * @param offsetX x offset for translation.
     * @param offsetY y offset for translation.
     */
    fun offset(offsetX: Double, offsetY: Double) {
        setLocation(locationX + offsetX, locationY + offsetY)
        events.updated.fire()
    }

    /**
     * Since Neuron Array is immutable, this object will be used in the creation dialog.
     */
    class CreationTemplate : EditableObject {
        /**
         * Size of the neuron array.
         */
        @UserParameter(label = "Nodes", description = "Number of nodes", order = 1)
        var numNodes = 100

        /**
         * Add a neuron array to network created from field values which should be setup by an Annotated Property
         * Editor.
         *
         * @param network the network this neuron array adds to
         * @return the created neuron array
         */
        fun create(): NeuronArray {
            return NeuronArray(numNodes)
        }

        override val name = "Neuron Array"

    }

    context(Network)
    override fun update() {
        if (isClamped) {
            return
        }
        updateRule.apply(this, dataHolder)
        inputs.mul(0.0) // clear inputs
        events.updated.fire()
    }

    @Consumable
    override fun setActivations(activations: DoubleArray) {
        this.activations = Matrix.column(activations)
    }

    /**
     * Set all activations in the array to the specified value.
     */
    fun fillActivations(value: Double) {
        this.activations.setColConstant(0, value)
    }


    fun fireLocationChange() {
        events.locationChanged.fire()
    }

    override fun toString(): String {
        return """
            $id with ${this.activations.size()} components.
            Activations: ${Utils.getTruncatedArrayString(activationArray, 10)}
            $dataHolder
        """.trimIndent()
    }

    override fun clear() {
        activations.setColConstant(0, 0.0)
        inputs.setColConstant(0, 0.0)
        events.updated.fire()
    }

    override fun increment() {
        activations.add(increment)
        events.updated.fire()
    }

    override fun decrement() {
        activations.sub(increment)
        events.updated.fire()
    }


    /**
     * See [org.simbrain.workspace.serialization.WorkspaceComponentDeserializer]
     */
    override fun readResolve(): Any? {
        events = NeuronArrayEvents()
        return this
    }

    /**
     * For each incoming psr matrix, filter out entries that correspond to excitatory synapses and return the sum of those for
     * each row (i.e. a vector of excitatory inputs; summed PSRs for each “dendrite”).
     */
    val excitatoryInputs: DoubleArray
        get() = incomingConnectors
            .filterIsInstance<WeightMatrix>()
            .map { it.psrMatrix.clone().mul(it.excitatoryMask).rowSums() }
            .reduceOrNull { base, add -> SimbrainMath.addVector(base, add) }
            ?: DoubleArray(size)

    /**
     * For each incoming psr matrix, filter out entries that correspond to inhibitory synapses and return the sum of those for
     * each row (i.e. a vector of inhibitory inputs; summed PSRs for each “dendrite”).
     */
    val inhibitoryInputs: DoubleArray
        get() = incomingConnectors
            .filterIsInstance<WeightMatrix>()
            .map { it.psrMatrix.clone().mul(it.inhibitoryMask).rowSums() }
            .reduceOrNull { base, add -> SimbrainMath.addVector(base, add) }
            ?: DoubleArray(size)

    fun getLabelArray() = (0 until size).map { it.toString() }.toTypedArray()
}